controller_connections.go

```diff
@@ -90,19 +92,116 @@ func (c *controller) handleIncoming(con net.Conn) {
 		return
 	}
 
-	peer := newPeer(c.net, c.peerStatus, c.peerData)
-	// we are never expecting a reject-alternate for incoming connections
-	if _, err := peer.StartWithHandshake(ep, con, true); err != nil {
-		c.logger.WithError(err).Debugf("Handshake failed for address %s, stopping", ep)
-		peer.Stop()
+	// don't bother with alternatives for incoming connections
+	peer, _, err := c.handshake(host, con, true)
+
+	if err != nil {
+		c.logger.WithError(err).Debugf("inbound connection from %s failed handshake", host)
+		con.Close()
 		return
 	}
 
 	c.logger.Debugf("Incoming handshake success for peer %s, version %s", peer.Hash, peer.prot.Version())
+}
+
+func (c *controller) handshake(ip string, con net.Conn, incoming bool) (*Peer, []Endpoint, error) {
+	tmplogger := c.logger.WithField("addr", ip)
+	timeout := time.Now().Add(c.net.conf.HandshakeTimeout)
+
+	nonce := make([]byte, 8) // loopback detection
+	binary.LittleEndian.PutUint64(nonce, c.net.instanceID)
+
+	// upgrade connection to a metrics connection
+	metrics := NewMetricsReadWriter(con)
+
+	handshake := newHandshake(c.net.conf, nonce)
+	decoder := gob.NewDecoder(metrics) // pipe gob through the metrics writer
+	encoder := gob.NewEncoder(metrics)
+	con.SetWriteDeadline(timeout)
+	con.SetReadDeadline(timeout)
+
+	failfunc := func(err error) (*Peer, []Endpoint, error) {
+		tmplogger.WithError(err).Debug("Handshake failed")
+		con.Close()
+		return nil, nil, err
+	}
+
+	err := encoder.Encode(handshake)
+	if err != nil {
+		return failfunc(fmt.Errorf("Failed to send handshake to incoming connection"))
```

Comment:
the usage of "incoming" in this log message is a bit confusing given that we have an `incoming` variable that can be true or false. Is it a different meaning for incoming here?

Reply:
I used "incoming" as a new peer coming into the system, which I agree is ambiguous. I dropped the phrasing and changed the logger to include the "incoming" bool as a field.